src/riak_kv_replrtq_peer.erl

```diff
@@ -0,0 +1,240 @@
+-spec update_discovery(riak_kv_replrtq_snk:queue_name()) -> boolean().
+update_discovery(QueueName) ->
+    gen_server:call(?MODULE, {update_discovery, QueueName}, 60 * 1000).
+
+-spec update_workers(pos_integer(), pos_integer()) -> boolean().
+update_workers(WorkerCount, PerPeerLimit) ->
+    gen_server:call(?MODULE,
+        {update_workers, WorkerCount, PerPeerLimit},
+        60 * 1000).
+
+-spec prompt_discovery(discovery_peer()) -> ok.
+prompt_discovery({QueueName, PeerInfo}) ->
+    gen_server:cast(?MODULE, {prompt_discovery, {QueueName, PeerInfo}}).
+
+
+%%%============================================================================
+%%% gen_server callbacks
+%%%============================================================================
+
+init([]) ->
+    case app_helper:get_env(riak_kv, replrtq_peer_discovery, false) of
+        true -> 
+            SinkPeers = app_helper:get_env(riak_kv, replrtq_sinkpeers, ""),
+            DefaultQueue = app_helper:get_env(riak_kv, replrtq_sinkqueue),
+            SnkQueuePeerInfo =
+                riak_kv_replrtq_snk:tokenise_peers(DefaultQueue, SinkPeers),
+
+            MinDelay = 
+                app_helper:get_env(riak_kv,
+                    replrtq_prompt_min_seconds,
+                    ?AUTO_DISCOVERY_MINIMUM_SECONDS),
+
+            lists:foreach(
+                fun({QueueName, _PeerInfo}) -> 
+                    erlang:send_after(MinDelay * 1000,
+                        self(),
+                        {prompt_discovery, QueueName})
+                end,
+                SnkQueuePeerInfo),
+            {ok, #state{discovery_peers = SnkQueuePeerInfo}};
+        false ->
+            {ok, #state{}}
+    end.
+
+handle_call({update_discovery, QueueName}, _From, State) ->
+    case lists:keyfind(QueueName, 1, State#state.discovery_peers) of
+        false ->
+            lager:info("Type=~w discovery for unconfigured QueueName=~w",
+                [update, QueueName]),
+            {reply, false, State};
+        {QueueName, PeerInfo} ->
+            R = prompt_discovery(QueueName, PeerInfo, update),
+            {reply, R, State}
+    end;
+handle_call({update_workers, WorkerCount, PerPeerLimit}, _From, State) ->
+    case riak_kv_replrtq_snk:get_worker_counts() of
+        {WorkerCount, PerPeerLimit} ->
+            {reply, false, State};
+        _ ->
+            riak_kv_replrtq_snk:set_worker_counts(WorkerCount, PerPeerLimit),
+            lists:foreach(
+                fun({QN, PI}) -> prompt_discovery(QN, PI, count_change) end,
+                State#state.discovery_peers),
+            {reply, true, State}
+    end.
+
+handle_cast({prompt_discovery, {QueueName, PeerInfo}}, State) ->
+    _ = prompt_discovery(QueueName, PeerInfo, regular),
+    {noreply, State}.
+
+handle_info({prompt_discovery, QueueName}, State) ->
+    {QueueName, PeerInfo} =
+        lists:keyfind(QueueName, 1, State#state.discovery_peers),
+    ok = prompt_discovery({QueueName, PeerInfo}),
```

Comment:
I kind of dislike this... you leave the context of the genserver in the cast to later return to it... I wonder if it would not be better to actually call `prompt_discovery(QueueName, PeerInfo, regular)` at this point.